src/asmcnc/job/job_data.py

```diff
@@ -389,3 +400,209 @@ def post_job_data_update_post_send(self):
+            print(str(traceback.format_exc()))
+
+
+    def reset_recovery(self):
+        self.job_recovery_selected_line = -1
+        self.job_recovery_gcode = []
+        self.job_recovery_offset = 0
+
+    def generate_recovery_gcode(self):
+        try:
+            recovery_gcode = []
+
+            # Recover modal gcodes
+
+            # Coordinate System Select
+            coord_system_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G5[4-9]", s)), None)
+            if coord_system_line:
+                recovery_gcode.append(re.search("G5[4-9](\.[1-3])?", coord_system_line).group(0))
+
+            # Plane selection
+            plane_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G1[7-9]", s)), None)
+            if plane_line:
+                recovery_gcode.append(re.search("G1[7-9]", plane_line).group(0))
+
+            # Distance mode
+            distance_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[0,1]([A-Z]|\s|$)", s)), None)
+            if distance_line:
+                # Recovery not allowed if G91 is used
+                if re.search("G91([A-Z]|\s|$)", distance_line):
+                    return False, 'This job contains the G91 g-code, so it cannot be recovered.'
+                recovery_gcode.append("G90")
+
+            # Arc IJK distance mode
+            arc_mode_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[0,1]\.1", s)), None)
+            if arc_mode_line:
+                recovery_gcode.append(re.search("G9[0,1]\.1", arc_mode_line).group(0))
+
+            # Feed rate mode
+            feedrate_mode_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[3-5]", s)), None)
+            if feedrate_mode_line:
+                recovery_gcode.append(re.search("G9[3-5]", feedrate_mode_line).group(0))
+
+            # Units
+            unit_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G2[0,1]", s)), None)
+            if unit_line:
+                recovery_gcode.append(re.search("G2[0,1]", unit_line).group(0))
+
+            # Cutter radius compensation
+            if next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if 'G40' in s), None):
+                recovery_gcode.append('G40')
+
+            # Tool length offset
+            tool_length_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G4(3\.1|9)", s)), None)
+            if tool_length_line:
+                recovery_gcode.append(re.search("G4(3\.1|9)", tool_length_line).group(0))
+
+            # Program mode
+            program_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("M0?[0-2](\D|$)|M30", s)), None)
```

Comment:
@BenjiPestell @benjipestell64 Can you do a test with X rack cutter file with pause to ensure this works in the way we want? Concerned that it will pause at job beginning, even if pause was cleared prior to cancellation.

Reply:
I had Dennis put in a fix for this, works fine now